org.springframework.beans.factory.BeanCreationException on hotswap

### Description of the bug

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'no.ka.studd.views.MainLayout': Failed to instantiate [no.ka.studd.views.MainLayout]: Constructor threw exception
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:321) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:309) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1377) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1214) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:563) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:523) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:322) ~[spring-beans-6.2.1.jar:6.2.1]
	at com.vaadin.flow.spring.SpringInstantiator.getOrCreate(SpringInstantiator.java:155) ~[vaadin-spring-24.7.0.alpha5.jar:na]
	at com.vaadin.flow.di.Instantiator.createRouteTarget(Instantiator.java:170) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.lambda$getRouteTarget$1(AbstractNavigationStateRenderer.java:146) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at java.base/java.util.Optional.orElseGet(Optional.java:364) ~[na:na]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.getRouteTarget(AbstractNavigationStateRenderer.java:145) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.sendBeforeEnterEventAndPopulateChain(AbstractNavigationStateRenderer.java:639) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.createChainIfEmptyAndExecuteBeforeEnterNavigation(AbstractNavigationStateRenderer.java:620) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.handleBeforeNavigationEvents(AbstractNavigationStateRenderer.java:326) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.handle(AbstractNavigationStateRenderer.java:206) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.Router.handleNavigation(Router.java:266) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.router.Router.navigate(Router.java:234) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.internal.UIInternals.refreshCurrentRoute(UIInternals.java:1133) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.UI.refreshCurrentRoute(UI.java:1259) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.hotswap.Hotswapper.lambda$triggerClientUpdate$295d7b78$1(Hotswapper.java:476) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.UI.accessSynchronously(UI.java:498) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.UI$2.execute(UI.java:573) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2173) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.server.VaadinSession.unlock(VaadinSession.java:755) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.server.VaadinService.ensureAccessQueuePurged(VaadinService.java:2137) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.server.VaadinService.accessSession(VaadinService.java:2104) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.server.VaadinSession.access(VaadinSession.java:1059) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.UI.access(UI.java:570) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.component.UI.access(UI.java:553) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.hotswap.Hotswapper.lambda$triggerClientUpdate$9(Hotswapper.java:476) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596) ~[na:na]
	at com.vaadin.flow.hotswap.Hotswapper.triggerClientUpdate(Hotswapper.java:476) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.hotswap.Hotswapper.onHotswapInternal(Hotswapper.java:266) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at com.vaadin.flow.hotswap.Hotswapper.onHotswap(Hotswapper.java:145) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.hotswap.agent.command.ReflectionCommand.doExecuteReflectionCommand(ReflectionCommand.java:207) ~[hotswap-agent.jar:2.0.1]
	at org.hotswap.agent.command.ReflectionCommand.executeCommand(ReflectionCommand.java:168) ~[hotswap-agent.jar:2.0.1]
	at org.hotswap.agent.command.impl.CommandExecutor.run(CommandExecutor.java:43) ~[hotswap-agent.jar:2.0.1]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [no.ka.studd.views.MainLayout]: Constructor threw exception
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:222) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:145) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:318) ~[spring-beans-6.2.1.jar:6.2.1]
	... 42 common frames omitted
Caused by: java.lang.IllegalStateException: No request is available. This method can only be used with an active VaadinServletRequest
	at com.vaadin.flow.server.auth.AccessAnnotationChecker.hasAccess(AccessAnnotationChecker.java:92) ~[flow-server-24.7.0.alpha5.jar:24.7.0.alpha5]
	at no.ka.studd.views.MainLayout.createNavigation(MainLayout.kt:105) ~[classes/:na]
	at no.ka.studd.views.MainLayout.addDrawerContent(MainLayout.kt:173) ~[classes/:na]
	at no.ka.studd.views.MainLayout.<init>(MainLayout.kt:99) ~[classes/:na]
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486) ~[na:na]
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Constructor.call(CallerImpl.kt:41) ~[kotlin-reflect-2.0.21.jar:2.0.21-release-482]
	at kotlin.reflect.jvm.internal.KCallableImpl.callDefaultMethod$kotlin_reflection(KCallableImpl.kt:207) ~[kotlin-reflect-2.0.21.jar:2.0.21-release-482]
	at kotlin.reflect.jvm.internal.KCallableImpl.callBy(KCallableImpl.kt:112) ~[kotlin-reflect-2.0.21.jar:2.0.21-release-482]
	at org.springframework.beans.BeanUtils$KotlinDelegate.instantiateClass(BeanUtils.java:940) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:190) ~[spring-beans-6.2.1.jar:6.2.1]
	... 44 common frames omitted


### Expected behavior

Works with HotswapAgent 1.4.2 and page refresh, not HotswapAgent 2.0.1.

**Should be an option to select a full page refresh in Copilot at runtime and a application.properties flag**

### Minimal reproducible example

Not sure why it fail, but Vaadin may understand looking at the stack trace

### Versions

Hilla: 24.7.0.alpha6
Flow: 24.7.0.alpha5
Vaadin: 24.7.0.alpha4
Copilot: 24.7.0.alpha4
Frontend Hotswap: Disabled, using pre-built bundle
OS: amd64 Windows 11 10.0
Java: JetBrains s.r.o. 21.0.5
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36
Java Hotswap: Java Hotswap is enabled
IDE Plugin: 1.4.5 IntelliJ

--- a/flow-server/src/main/java/com/vaadin/flow/hotswap/Hotswapper.java
+++ b/flow-server/src/main/java/com/vaadin/flow/hotswap/Hotswapper.java
@@ -472,9 +472,11 @@ public class Hotswapper implements ServiceDestroyListener, SessionInitListener,
             LOGGER.debug(
                     "Triggering re-navigation to current route for UIs affected by classes changes.");
             for (UIRefreshStrategy action : uisToRefresh.keySet()) {
-                uisToRefresh.get(action)
-                        .forEach(ui -> ui.access(() -> ui.refreshCurrentRoute(
-                                action == UIRefreshStrategy.PUSH_REFRESH_CHAIN)));
+                String triggerEventJS = String.format(
+                        "window.dispatchEvent(new CustomEvent(\"vaadin-refresh-ui\", { detail: { fullRefresh: %s }}));",
+                        action == UIRefreshStrategy.PUSH_REFRESH_CHAIN);
+                uisToRefresh.get(action).forEach(ui -> ui
+                        .access(() -> ui.getPage().executeJs(triggerEventJS)));
             }
         }
     }
@@ -507,7 +509,19 @@ public class Hotswapper implements ServiceDestroyListener, SessionInitListener,
 
     @Override
     public void uiInit(UIInitEvent event) {
-        sessions.add(event.getUI().getSession());
+        UI ui = event.getUI();
+        sessions.add(ui.getSession());
+        ui.getPage().executeJs(
+                """
+                        const $wnd = window;
+                        window.addEventListener('vaadin-ui-refresh', (ev) => {
+                            const senderFn = $wnd.Vaadin?.Flow?.clients[$0]?.sendEventMessage;
+                            if (senderFn) {
+                                senderFn(1, "ui-refresh", ev.detail);
+                            }
+                        });
+                        """,
+                ui.getInternals().getAppId());
     }
 
     /**

--- a/flow-server/src/test/java/com/vaadin/flow/hotswap/HotswapperTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/hotswap/HotswapperTest.java
@@ -24,6 +24,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import org.junit.Assert;
 import org.junit.Before;
@@ -35,6 +37,7 @@ import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.HasComponents;
 import com.vaadin.flow.component.HasElement;
 import com.vaadin.flow.component.Tag;
+import com.vaadin.flow.component.page.Page;
 import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.internal.BrowserLiveReload;
 import com.vaadin.flow.internal.BrowserLiveReloadAccessor;
@@ -1153,20 +1156,30 @@ public class HotswapperTest {
 
     private static class RefreshTestingUI extends MockUI {
 
+        private static final Pattern UI_REFRESH_EVENT = Pattern.compile(
+                ".*new CustomEvent\\(\"vaadin-refresh-ui\",\\s*\\{\\s*detail:\\s*\\{\\s*fullRefresh:\\s*(true|false)\\s*}\\s*}\\).*");
         private Boolean refreshRouteChainRequested;
+
+        private final Page pageSpy;
 
         public RefreshTestingUI(VaadinSession session) {
             super(session);
+            pageSpy = Mockito.spy(super.getPage());
+            // Intercept javascript executions to check if the custom ui refresh
+            // event dispatch has been registered.
+            Mockito.doAnswer(i -> {
+                Matcher matcher = UI_REFRESH_EVENT.matcher(i.getArgument(0));
+                if (matcher.matches()) {
+                    refreshRouteChainRequested = Boolean
+                            .parseBoolean(matcher.group(1));
+                }
+                return null;
+            }).when(pageSpy).executeJs(Mockito.anyString());
         }
 
         @Override
-        public void refreshCurrentRoute(boolean refreshRouteChain) {
-            refreshRouteChainRequested = refreshRouteChain;
-            // No need to perform real navigation, tests only need to know if
-            // the method has been invoked.
-            // Navigation would fail anyway because of usage of method scoped
-            // classes. Blocking navigation prevents logs to be bloated by
-            // exception stack traces.
+        public Page getPage() {
+            return pageSpy;
         }
 
         void assertNotRefreshed() {

